---
title: "IBD prediction"
output: html_notebook
---

```{r}
library(data.table)
library(readxl)
library(ggplot2)
library(gprofiler2)
library(glmnet)
library(ComplexHeatmap)
set.seed(101)
```
# Data preprocessing
### Read data
```{r}
# read meta data
meta <- as.data.table(readxl::read_excel("data/transcritome_patients_translated.xlsx"))

meta_key <- "CEL_FILE"
setkeyv(meta, meta_key)                             # set key
stopifnot(!any( duplicated(meta[,..meta_key]) ))    # check for duplicate rows
stopifnot(!any( duplicated(colnames(meta)) ))       # check for duplicate columns


# read expression data
expr <- as.data.table(readxl::read_excel("data/transcriptome_expression_matrix.xlsx"))
expr_key <- "gene"
colnames(expr)[1] <- expr_key
setkeyv(expr, expr_key)                             # set key
stopifnot(!any( duplicated(expr[,..expr_key]) ))    # check for duplicate rows
stopifnot(!any( duplicated(colnames(expr)) ))       # check for duplicate columns

expr <- expr[, c(key(expr), meta$CEL_FILE), with=F]
stopifnot(all(colnames(expr)[-1]==meta[,CEL_FILE]))
expr <- as.matrix(expr, rownames="gene")

meta_M0 <- meta[LOCATION=="M0"]
meta_MI <- meta[LOCATION=="MI"]
meta_M6 <- meta[LOCATION=="M6"]

# read mouse signature
signature_mouse <- fread("data/signature.csv",header = F)[[1]]

# read all mouse genes measured with nanostring 
measured_genes_mouse <- fread("data/All samples_NormalizedData.csv")[[1]]

# check that all signature genes are part of the measured genes
stopifnot(all(signature_mouse %in% measured_genes_mouse))
```
### Ortholog mapping
```{r}
# map to human orthologs
ortholog_mapping <- as.data.table(gprofiler2::gorth(measured_genes_mouse, source_organism = "mmusculus", target_organism = "hsapiens", filter_na = F))
stopifnot(all(measured_genes_mouse %in% ortholog_mapping$input))

# mark signature genes and hits in the human expr data
ortholog_mapping$in_expr <- ortholog_mapping$ortholog_name %in% rownames(expr)
ortholog_mapping$in_signature <- ortholog_mapping$input %in% signature_mouse 

# count number of hits in the human expr data per mouse gene
n_in_expr <- ortholog_mapping[,.("n_hits" = sum(in_expr)),by=input]

# print number of hits in expr$gene per measured mouse gene
table(n_in_expr$n_hits)
```

```{r}
# select only mouse genes with orthologs uniquely mapped to the human expression data
unique_ortholog_mapping <- ortholog_mapping[input %in% n_in_expr[n_hits==1,input] & in_expr]
# unmapped signature genes
print(paste0(nrow(unique_ortholog_mapping)," out of ",length(measured_genes_mouse), " measured mouse genes were mapped to the human expression genes"))
ortholog_mapping[!input %in% unique_ortholog_mapping$input & in_signature]
```

# Data checks
#### Expression sum per sample
```{r}
plot_data <- data.table(sample=colnames(expr))
plot_data[,colSum:=colSums(expr)]
plot_data[,RECURRENCE:=meta$RECURRENCE]
plot_data[,LOCATION:=meta$LOCATION]
plot_data[,GENDER:=meta$GENDER]

ggplot(plot_data,aes(x=colSum,fill=RECURRENCE)) +
  geom_density(alpha=0.5)

ggplot(plot_data,aes(x=colSum,fill=LOCATION)) +
  geom_density(alpha=0.5)

ggplot(plot_data,aes(x=colSum,fill=GENDER)) +
  geom_density(alpha=0.5)
```
#### Outcome distribution
```{r}
table(meta$RECURRENCE, meta$LOCATION)
```
#### Signature heatmap
```{r}
column_ha = HeatmapAnnotation(GENDER = meta$GENDER, LOCATION = meta$LOCATION, RECURRECE=meta$RECURRENCE)
Heatmap(t(apply( expr[unique_ortholog_mapping[in_expr==T&in_signature==T,ortholog_name],meta$CEL_FILE] ,1,scale)),
        show_column_names = F, 
        top_annotation = column_ha,
        name = "Expression",
        )

column_ha = HeatmapAnnotation(GENDER = meta_M0$GENDER, LOCATION = meta_M0$LOCATION, RECURRECE=meta_M0$RECURRENCE)
Heatmap(t(apply( expr[unique_ortholog_mapping[in_expr==T&in_signature==T,ortholog_name],meta_M0$CEL_FILE] ,1,scale)),
        show_column_names = F, 
        top_annotation = column_ha,
        name = "Expression",
        )

column_ha = HeatmapAnnotation(GENDER = meta_M6$GENDER, LOCATION = meta_M6$LOCATION, RECURRECE=meta_M6$RECURRENCE)
Heatmap(t(apply( expr[unique_ortholog_mapping[in_expr==T&in_signature==T,ortholog_name],meta_M6$CEL_FILE] ,1,scale)),
        show_column_names = F, 
        top_annotation = column_ha,
        name = "Expression",
        )

column_ha = HeatmapAnnotation(GENDER = meta_MI$GENDER, LOCATION = meta_MI$LOCATION, RECURRECE=meta_MI$RECURRENCE)
Heatmap(t(apply( expr[unique_ortholog_mapping[in_expr==T&in_signature==T,ortholog_name],meta_MI$CEL_FILE] ,1,scale)),
        show_column_names = F, 
        top_annotation = column_ha,
        name = "Expression",
        )
```

#### High variance genes heatmap
```{r}
n_genes <- length(unique_ortholog_mapping[in_expr==T&in_signature==T,ortholog_name])
high_var_genes <- sort(apply(expr, 1, sd), decreasing = T)[1:n_genes]

column_ha = HeatmapAnnotation(GENDER = meta$GENDER, LOCATION = meta$LOCATION, RECURRECE=meta$RECURRENCE)
Heatmap(t(apply( expr[names(high_var_genes),meta$CEL_FILE] ,1,scale)),
        show_column_names = F, 
        top_annotation = column_ha,
        name = "Expression",
        )

column_ha = HeatmapAnnotation(GENDER = meta_M0$GENDER, LOCATION = meta_M0$LOCATION, RECURRECE=meta_M0$RECURRENCE)
Heatmap(t(apply( expr[names(high_var_genes),meta_M0$CEL_FILE] ,1,scale)),
        show_column_names = F, 
        top_annotation = column_ha,
        name = "Expression",
        )

column_ha = HeatmapAnnotation(GENDER = meta_M6$GENDER, LOCATION = meta_M6$LOCATION, RECURRECE=meta_M6$RECURRENCE)
Heatmap(t(apply( expr[names(high_var_genes),meta_M6$CEL_FILE] ,1,scale)),
        show_column_names = F, 
        top_annotation = column_ha,
        name = "Expression",
        )

column_ha = HeatmapAnnotation(GENDER = meta_MI$GENDER, LOCATION = meta_MI$LOCATION, RECURRECE=meta_MI$RECURRENCE)
Heatmap(t(apply( expr[names(high_var_genes),meta_MI$CEL_FILE] ,1,scale)),
        show_column_names = F, 
        top_annotation = column_ha,
        name = "Expression",
        )
```
# Recurrence prediction
## Feature selection with logistic glm-net (M0)
```{r}
x_genes <- unique_ortholog_mapping[in_signature==T,ortholog_name]
x <- t(expr[x_genes,meta_M0$CEL_FILE])
y <- meta_M0$RECURRENCE

fit <- glmnet(x, y, family = "binomial")
plot(fit, label = T)
cvfit <- cv.glmnet(x, y, family = "binomial", type.measure = "class")
plot(cvfit)
print(cvfit)

data <- as.data.frame(x)
data$RECURRENCE <- 0
data$RECURRENCE[y=="R"] <- 1
glm_model <- glm(RECURRENCE ~.,family=binomial(link='logit'), data)
```
## Feature selection with logistic glm-net (M6)
```{r}
x_genes <- unique_ortholog_mapping[in_signature==T,ortholog_name]
x <- t(expr[x_genes,meta_M6$CEL_FILE])
y <- meta_M6$RECURRENCE

fit <- glmnet(x, y, family = "binomial")
plot(fit, label = T)
cvfit <- cv.glmnet(x, y, family = "binomial", type.measure = "class")
plot(cvfit)
print(cvfit)
```
## Feature selection with logistic glm-net (MI)
```{r}
x_genes <- unique_ortholog_mapping[in_signature==T,ortholog_name]
x <- t(expr[x_genes,meta_MI$CEL_FILE])
y <- meta_MI$RECURRENCE

fit <- glmnet(x, y, family = "binomial")
plot(fit, label = T)
cvfit <- cv.glmnet(x, y, family = "binomial", type.measure = "class")
plot(cvfit)
print(cvfit)
```
